File: src/storeFx/goToNextSong.ts
import * as Fx from '../fx/fx';
import type { Store } from '../store/types';
import * as StoreFx__StartPlaying from './startPlaying';

export function goToNextSong() {
  return Fx.bind(Fx.ask<{ store: Store }>(), ({ store }) => {
    const state = store.getState();
    const queued = state.player.queuedSongIds;

    const currentIndex =
      state.player.currentSongId == null
        ? 0
        : queued.indexOf(state.player.currentSongId);

    return StoreFx__StartPlaying.make(
      queued[(currentIndex + 1) % queued.length],
    );
  });
}

File: src/storeFx/goToNextSong.ml
open Fx
open Monad_syntax

let make () =
  let* Deps.{ store } = Fx.ask () in
  let state = store |. Zustand.getState in
  let queued = state.player.queuedSongIds in
  let len = queued |> Js.Array.length
  and currentIndex =
    state.player.currentSongId
    |> Option.map (fun value -> queued |> Js.Array.indexOf ~value)
    |> Option.value ~default:0
  in
  let current = queued |. Belt.Array.getUnsafe ((currentIndex + 1) mod len) in
  StartPlaying.make ~current ()
